contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -885,6 +912,164 @@ run_tsql_best_match_heuristics(int nargs, Oid *input_typeids, FuncCandidateList
+	 */
+	if (func_nsname != NULL &&
+		(strlen(func_nsname) != 3 || strncmp(func_nsname, "sys", 3) != 0))
+		return false;
+
+	/* Get Exception function details */
+	for (i = 0; i < TOTAL_TSQL_EXCEPTION_FUNCTION_COUNT; i++)
+	{
+		exception_func = &tsql_exception_function_list[i];
+		if (strcmp(func_name, exception_func->funcname) == 0
+			&& nargs == exception_func->nargs)
+				break;
+		exception_func = NULL;
+	}
+
+	/* If function is not an exception function no additional handling required */
+	if (i >= TOTAL_TSQL_EXCEPTION_FUNCTION_COUNT
+		|| exception_func == NULL)
+	{
+		return false;
+	}
+
+	sys_varcharoid = get_sys_varcharoid();
+	/* Report error in case of invalid argument datatype */
+	for (i = 0; i < exception_func->nargs; i++)
+	{
+		/* for unknown literals consider its type as sys.VARCHAR */
+		inputTypId = (input_typeids[i] == UNKNOWNOID) ? sys_varcharoid : input_typeids[i];
+
+		type_match = false;
+		for (int j = 0; j < exception_func->valid_arg_types[i].len; j++)
+		{
+			if (!OidIsValid(exception_func->valid_arg_types[i].valid_types_oid[j]))
+				exception_func->valid_arg_types[i].valid_types_oid[j] = 
+					(*common_utility_plugin_ptr->lookup_tsql_datatype_oid) (exception_func->valid_arg_types[i].valid_types[j]);
+
+			validTypId = exception_func->valid_arg_types[i].valid_types_oid[j];
+
+			if (inputTypId == validTypId)
+			{
+				type_match = true;
+				break;
+			}
+		}
+		if (!type_match)
+		{
+			ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_FUNCTION),
+				 errmsg("Argument data type %s is invalid for argument %d of %s function.", 
+				 		 format_type_be(input_typeids[i]), i+1, exception_func->funcname)));
+		}
+	}
+
+	return true;
+}
+
+/*
+ * tsql_func_select_candidate_for_exception()
+ *
+ * For functions present in exception function list, and try to find best candidate 
+ * based on matching return type. Also throw error in case of invalid argument data type.
+ */
+static FuncCandidateList
+tsql_func_select_candidate_for_exception(List *names, int nargs, Oid *input_typeids, FuncCandidateList candidates)
+{
+	FuncCandidateList			current_candidate, best_candidate;
+	HeapTuple					tuple;
+	Oid 						expr_result_type;
+	Form_pg_proc 				procform;
+	char					   *proc_nsname;
+	char					   *proc_name;
+	bool						is_func_validated;
+	int							ncandidates;
+
+	if (nargs > FUNC_MAX_ARGS)
+		ereport(ERROR,
+				(errcode(ERRCODE_TOO_MANY_ARGUMENTS),
+				 errmsg_plural("cannot pass more than %d argument to a function",
+								"cannot pass more than %d arguments to a function",
+								FUNC_MAX_ARGS,
+								FUNC_MAX_ARGS)));
+
+	DeconstructQualifiedName(names, &proc_nsname, &proc_name);
+
+	is_func_validated = validate_exception_function(proc_nsname, proc_name, nargs, input_typeids);
+
+	/* Return NULL if function is not an exception function */
+	if (!is_func_validated)
+		return NULL;
+
+	/* function based logic to decide return type */
+	expr_result_type = -1;
+	if (strcmp(proc_name,"trim") == 0)
```

Comment:
should check sys schmea also

Reply:
schema name is already been validated in `validate_exception_function` function